### PR TITLE
Fix/override kwargs base block

### PR DESCRIPTION
**Pull request by mihirathale98**
This PR enables runtime parameter override functionality for all blocks in the SDG Hub framework, providing flexible block configuration without modifying YAML files.

 #### Features

  - Runtime field overrides for any block type during execution
  - Full Pydantic validation of override values with proper error messages
  - Automatic value restoration after block execution (exception-safe)
  - Enhanced TextParserBlock validation for reasoning content fields
  - Zero breaking changes - existing code continues to work unchanged

  #### Benefits

  - Eliminates(still supoorts) manual YAML editing for runtime parameter changes
  - Enables dynamic block configuration in flows and pipelines
  - Improves developer experience with validated, type-safe overrides
  - Supports experimentation with different block configurations

  #### Usage

```python


  # In flows (via runtime_params)
  flow_path = FlowRegistry.get_flow_path('small-rock-799')
  flow = Flow.from_yaml(flow_path)

  runtime_params = {
      "parse_atomic_facts": {
          "save_reasoning_content": True,
      }
  }

  result = flow.generate(dataset, runtime_params=runtime_params)
```
  #### Files Changed

  - src/sdg_hub/core/blocks/base.py - Added kwargs override system to BaseBlock
  - src/sdg_hub/core/blocks/llm/text_parser_block.py - Added reasoning content validation
  - tests/blocks/test_base_block.py - Tests for BaseBlock kwargs override functionality
  - tests/blocks/llm/test_textparserblock.py - Tests for reasoning content validation

  #### Summary

  New Features:
  - Runtime parameter override for all blocks with Pydantic validation
  - Enhanced TextParserBlock with reasoning content field validation
  - Comprehensive error handling and automatic value restoration

  Tests:
  - 11 tests for BaseBlock kwargs override functionality
  - 7 tests for TextParserBlock reasoning content validation
  - Full coverage of validation, edge cases, and integration scenarios

## Summary by CodeRabbit

- **New Features**
  - Per-call runtime overrides with validation and guaranteed restoration; flexible per-call LLM parameter forwarding (llm_kwargs) and improved async concurrency control; wrappers now explicitly route and prioritize parameters to internal blocks.

- **Bug Fixes**
  - Validation for reasoning content field (non-empty + collision checks) and validation of merged configuration before applying overrides.

- **Refactor**
  - Removed legacy LLM config/client manager and simplified LLM wiring and public exports.

- **Tests**
  - Expanded/updated tests for overrides, parsing, LLM flows, concurrency, and validation.

### CODE REVIEW

**Comment by shivchander** [comment]
@codex review

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 1]
```diff
@@ -224,83 +145,36 @@ def model_post_init(self, __context) -> None:
...
                     "block_name": self.block_name,
                     "model": self.model,
-                    "provider": self.client_manager.config.get_provider(),
-                    "is_local": self.client_manager.config.is_local_model(),
                     "async_mode": self.async_mode,
-                    "generation_params": self.client_manager.config.get_generation_kwargs(),
                 },
             )
 
     def _reinitialize_client_manager(self) -> None:
```
Comment (eshwarprasadS) [inline src/sdg_hub/core/blocks/llm/llm_chat_block.py]: should we remove this method?

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 2]
```diff
@@ -383,54 +247,163 @@ def generate(self, samples: Dataset, **override_kwargs: dict[str, Any]) -> Datas
...
+        # Start with extra fields (temperature, max_tokens, etc.) from extra="allow"
+        extra_values = self.model_dump(exclude_unset=True)
+
+        # Remove block-operational fields that shouldn't go to LiteLLM
+        block_only_fields = {
+            "block_name",
+            "input_cols",
+            "output_cols",
+            "async_mode",
+            "max_retries",
```
Comment (eshwarprasadS) [inline src/sdg_hub/core/blocks/llm/llm_chat_block.py]: can we check if `max_retries` is part of the litellm args and not a block_only arg?

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 3]
```diff
@@ -383,54 +247,163 @@ def generate(self, samples: Dataset, **override_kwargs: dict[str, Any]) -> Datas
...
+            completion_kwargs["api_base"] = self.api_base
+        if self.timeout is not None:
+            completion_kwargs["timeout"] = self.timeout
+
+        # Apply runtime overrides (from BaseBlock + Flow)
+        completion_kwargs.update(overrides)
+
+        return completion_kwargs
+
+    def _completion_with_retry(self, messages, completion_kwargs):
```
Comment (eshwarprasadS) [inline src/sdg_hub/core/blocks/llm/llm_chat_block.py]: we should remove re-implementing retry mechanisms and defer to litellm

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 4]
```diff
@@ -383,54 +247,163 @@ def generate(self, samples: Dataset, **override_kwargs: dict[str, Any]) -> Datas
...
+                logger.warning(
+                    f"Retryable error on attempt {attempt + 1}/{max_retries + 1}: {str(e)}. "
+                    f"Retrying in {wait_time} seconds..."
+                )
+                time.sleep(wait_time)
+            except Exception as e:
+                # Non-retryable error, raise immediately
+                raise
+
+    async def _acompletion_with_retry(self, messages, completion_kwargs):
```
Comment (eshwarprasadS) [inline src/sdg_hub/core/blocks/llm/llm_chat_block.py]: same as above

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 5]
```diff
@@ -383,54 +247,163 @@ def generate(self, samples: Dataset, **override_kwargs: dict[str, Any]) -> Datas
...
+                logger.warning(
+                    f"Retryable error on attempt {attempt + 1}/{max_retries + 1}: {str(e)}. "
+                    f"Retrying in {wait_time} seconds..."
+                )
+                await asyncio.sleep(wait_time)
+            except Exception as e:
+                # Non-retryable error, raise immediately
+                raise
+
+    def _extract_response(self, response) -> Union[dict, list[dict]]:
```
Comment (eshwarprasadS) [inline src/sdg_hub/core/blocks/llm/llm_chat_block.py]: we should return the entire response dict as is from llmchat block

**Comment by shivchander** [comment]
@codex review

File: src/sdg_hub/core/blocks/base.py [hunk 6]
```diff
@@ -261,18 +261,73 @@ def __call__(self, samples: Dataset, **kwargs: Any) -> Dataset:
...
+                    self.__class__.model_validate(merged_config)
+                except Exception as e:
+                    raise ValueError(
+                        f"Invalid runtime override for {self.__class__.__name__}: {e}"
+                    ) from e
+
+                # Apply temporary overrides for block fields
+                for key, value in block_overrides.items():
+                    original_values[key] = getattr(self, key)
+                    setattr(self, key, value)
```
[the remark this answers is not in the document]
Comment (eshwarprasadS) [reply src/sdg_hub/core/blocks/base.py:309]: @mihirathale98 This is true, we are validating and might apply some type coercions with Pydantic that are being thrown away (the validated_model), and we might want to keep it. For eg:


```
llm_block = LLMChatBlock(
      block_name="test",
      model="gpt-4",
      temperature=0.5,     # float
      max_tokens=100       # int  
  )
```
if we runtime override the llm_block with strings instead of float / int: 
  
```
llm_block(dataset, 
temperature="0.9", 
max_tokens="200")
```
                        
the validate_model gets called but it's work will be thrown away and things that might expect temperate to be `int` might complain / break later in flow.

File: src/sdg_hub/core/blocks/llm/llm_chat_block.py [hunk 7]
```diff
@@ -2,21 +2,23 @@
...
 from ...utils.error_handling import BlockValidationError
 from ...utils.logger_config import setup_logger
+
+# Local
 from ..base import BaseBlock
 from ..registry import BlockRegistry
-from .client_manager import LLMClientManager
-from .config import LLMConfig
 
+litellm.drop_params = True
```
[the remark this answers is not in the document]
Comment (shivchander) [reply src/sdg_hub/core/blocks/llm/llm_chat_block.py:21]: @eshwarprasadS @mihirathale98, lets be careful with this

File: src/sdg_hub/core/blocks/evaluation/evaluate_relevancy_block.py [hunk 8]
```diff
@@ -141,41 +141,60 @@ def __init__(self, **kwargs):
...
+            params = self._add_composite_fields(params, block_class, wrapper_params)
+            params = {k: v for k, v in params.items() if k not in remove_params}
+        else:
+            # For other blocks, only forward parameters they accept
+            params = {
+                k: v
+                for k, v in kwargs.items()
+                if k in block_class.model_fields and k not in wrapper_params
+            }
+            params = self._add_composite_fields(params, block_class, wrapper_params)
```
[the remark this answers is not in the document]
Comment (shivchander) [reply src/sdg_hub/core/blocks/evaluation/evaluate_relevancy_block.py:183]: @eshwarprasadS could you take a look at this
